batch/batch/driver/instance_collection/pool.py

```diff
@@ -334,25 +350,42 @@ async def ready_cores_mcpu_from_estimated_job_queue(self):
     LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
   )
 ) AS t
+ORDER BY batch_id ASC, always_run DESC, job_id ASC
 LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
 )
 '''
 
             jobs_query.append(user_job_query)
             jobs_query_args += [user, self.name, user, self.name]
 
-        result = await self.db.select_and_fetchone(
+        result = self.db.select_and_fetchall(
+            # We use the string representation of regions JSON_UNQUOTE for order by instead of the json array
             f'''
-SELECT CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
+WITH ready_jobs AS (
+    {" UNION ".join(jobs_query)}
+)
+SELECT regions, CAST(COALESCE(SUM(cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
 FROM (
-{" UNION ".join(jobs_query)}
-) AS ready_jobs
+  SELECT regions, cores_mcpu,
+    CAST(ROW_NUMBER() OVER (ORDER BY scheduling_iteration, user, ready_jobs.batch_id, always_run DESC, -n_regions DESC, JSON_UNQUOTE(regions), ready_jobs.job_id) AS SIGNED) AS rn1,
+    CAST(ROW_NUMBER() OVER (PARTITION BY JSON_UNQUOTE(regions) ORDER BY scheduling_iteration, user, ready_jobs.batch_id, always_run DESC, -regions.n_regions DESC, JSON_UNQUOTE(regions.regions), ready_jobs.job_id) AS SIGNED) AS rn2
+  FROM ready_jobs
+  LEFT JOIN (
+    SELECT ready_jobs.batch_id, ready_jobs.job_id, JSON_ARRAYAGG(region) AS regions, COUNT(region) AS n_regions
+    FROM ready_jobs
+    INNER JOIN job_regions ON ready_jobs.batch_id = job_regions.batch_id AND ready_jobs.job_id = job_regions.job_id
+    LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id
+    GROUP BY ready_jobs.batch_id, ready_jobs.job_id
+  ) AS regions ON ready_jobs.batch_id = regions.batch_id AND ready_jobs.job_id = regions.job_id
+) AS ordered_jobs
+GROUP BY regions, rn1 - rn2
+ORDER BY rn1;
```

Comment:
What's the limit at the end of the day here? How many rows are we expected to return?

Reply:
The number of rows expected can be anywhere from 1 (I filtered out records with ready_cores_mcpu = 0 now) to the number of jobs that we're considering in the head of the queue. I put a limit of `MAX_INSTANCES_PER_AUTOSCALER_LOOP * self.worker_cores` now which I think should be plenty. So that's 160 now. In most cases, we should get 1 to 2 records. The only way you get lots of records is if users submit a bunch of jobs that have random regions required so the scan is not continuous.

Reply:
Sounds good, seems like a valid limit.